### PR TITLE
Add types to ZodContext.invalid

### PR DESCRIPTION
**Pull request by tomaskallup**
I've been trying to use this router in a project of mine, where I would like to handle validation my way (I return html `<form>` and I want to embed the errors in it) and ran into issue with types for the `ctx.invalid`. Since it didn't give any type to `ZodError`, it had no properties and I had to force them via type assertion.

This PR tries to fix that, I hope I did it right. Everything seems to build, tests do pass.

### CODE REVIEW

**Comment by JakeFenley** [comment]
hi @tomaskallup thank you for this, this looks great! I'll have a closer look at this when i get a chance

File: src/validation-middleware.ts [hunk 1]
```diff
@@ -90,13 +94,13 @@ export const validationMiddleware = <H, P, Q, B, F, R>(
...
 
-    const output = await validate(ctx.body, validation.response, 'response');
+    const output = await validate<R>(ctx.body, validation.response, 'response');
 
     if (!output) {
       return;
     }
 
-    if (output.error instanceof ZodError) {
+    if (assertZodValidationError<R>(output)) {
```
Comment (JakeFenley) [inline src/validation-middleware.ts:103]: is this change needed here, it doesn't seem to relate to adding the errors to ctx? is the type no longer a ZodError?

File: src/validation-middleware.ts [hunk 2]
```diff
@@ -47,19 +47,23 @@ export const validationMiddleware = <H, P, Q, B, F, R>(
...
 
-    const inputErrors = validated.reduce((acc: ZodRouterInvalid, curr) => {
-      if (assertZodValidationError(curr)) {
-        acc[curr.requestParameter] = curr.error;
-      }
-      return acc;
-    }, {});
+    const inputErrors = validated.reduce<ZodRouterInvalid<H, P, Q, B, F>>(
+      (acc, curr) => {
+        if (assertZodValidationError<any>(curr)) {
```
Comment (JakeFenley) [inline src/validation-middleware.ts:59]: why do we need <any> here?

File: src/validation-middleware.ts (hunk 1 above)
Comment (tomaskallup) [reply src/validation-middleware.ts:103]: It's needed to allow the generic type `<R>`, seems like that cannot be done with `instanceof ZodError<R>`, furthermore, the `assertZodValidationError` does the same, so I felt like it would be much better to unify this check.

File: src/validation-middleware.ts (hunk 2 above)
Comment (tomaskallup) [reply src/validation-middleware.ts:59]: Oh, we don't, it's a leftover when I was trying to get Typescript to cooperate, I'll remove it.

File: src/validation-middleware.ts (hunk 1 above)
Comment (JakeFenley) [reply src/validation-middleware.ts:103]: makes sense, its been awhile since i've worked on this :) 

thanks Tomas!
